packages/manager/apps/pci-instances/src/data/hooks/instance/selectors/instances.selector.ts

```diff
@@ -0,0 +1,159 @@
+import { InfiniteData } from '@tanstack/react-query';
+import { TInstanceDto, TInstanceStatusDto } from '@/types/instance/api.type';
+import {
+  TAddress,
+  TInstance,
+  TInstanceAction,
+  TInstanceActions,
+  TInstanceAddressType,
+  TInstanceStatus,
+  TInstanceStatusSeverity,
+} from '@/types/instance/entity.type';
+import { TActionName } from '@/types/instance/common.type';
+
+const getInstanceStatusSeverity = (
+  status: TInstanceStatusDto,
+): TInstanceStatusSeverity => {
+  switch (status) {
+    case 'BUILDING':
+    case 'REBOOT':
+    case 'REBUILD':
+    case 'REVERT_RESIZE':
+    case 'SOFT_DELETED':
+    case 'VERIFY_RESIZE':
+    case 'MIGRATING':
+    case 'RESIZE':
+    case 'BUILD':
+    case 'SHUTOFF':
+    case 'RESCUE':
+    case 'SHELVED':
+    case 'SHELVED_OFFLOADED':
+    case 'RESCUING':
+    case 'UNRESCUING':
+    case 'SNAPSHOTTING':
+    case 'RESUMING':
+    case 'HARD_REBOOT':
+    case 'PASSWORD':
+    case 'PAUSED':
+      return 'warning';
+    case 'DELETED':
+    case 'ERROR':
+    case 'STOPPED':
+    case 'SUSPENDED':
+    case 'UNKNOWN':
+      return 'error';
+    case 'ACTIVE':
+    case 'RESCUED':
+    case 'RESIZED':
+      return 'success';
+    default:
+      return 'info';
+  }
+};
+
+const getInstanceStatus = (status: TInstanceStatusDto): TInstanceStatus => ({
+  state: status,
+  severity: getInstanceStatusSeverity(status),
+});
+
+const getActionHrefByName = (
+  projectUrl: string,
+  name: TActionName,
+  { region, id }: Pick<TInstance, 'id' | 'region'>,
+): TInstanceAction['link'] => {
+  if (name === 'details') {
+    return { path: id, isExternal: false };
+  }
+
+  if (name === 'create_autobackup') {
+    return { path: `${projectUrl}/workflow/new`, isExternal: true };
+  }
+
+  if (name === 'assign_floating_ip') {
+    const searchParams = new URLSearchParams({
+      ipType: 'floating_ip',
+      region,
+      instance: id,
+    });
+
+    return {
+      path: `${projectUrl}/public-ips/order?${searchParams.toString()}`,
+      isExternal: true,
+    };
+  }
+
+  if (name === 'soft_reboot') {
+    return {
+      path: `region/${region}/instance/${id}/soft-reboot`,
+      isExternal: false,
+    };
+  }
+
+  if (name === 'hard_reboot') {
+    return {
+      path: `region/${region}/instance/${id}/hard-reboot`,
+      isExternal: false,
+    };
+  }
+
+  const actions = new Set([
+    'delete',
+    'stop',
+    'start',
+    'shelve',
+    'unshelve',
+    'reinstall',
+  ]);
+
+  if (actions.has(name)) {
+    return {
+      path: `region/${region}/instance/${id}/${name}`,
+      isExternal: false,
+    };
+  }
+
+  return { path: '', isExternal: false };
+};
+
+const mapInstanceAddresses = (instance: TInstanceDto) =>
+  instance.addresses.reduce((acc, { type, ...rest }) => {
+    if (acc.has(type)) {
+      const foundAddresses = acc.get(type) ?? [];
```

Comment:
```suggestion
      const foundAddresses = acc.get(type)!;
```
Here you know that it has been found

Reply:
It throws a forbidden non-null assertion warning, so we could keep the if predicate